pysph/base/kernels.py

```diff
@@ -89,31 +89,41 @@ def kernel(self, xij=[0., 0, 0], rij=1.0, h=1.0):
 
         return val * fac
 
-    def gradient(self, xij=[0., 0, 0], rij=1.0, h=1.0, grad=[0, 0, 0]):
+    def sigmadwdq(self, rij=1.0, h=1.0):
         h1 = 1. / h
         q = rij * h1
 
-        # get the kernel normalizing factor
+        # get the kernel normalizing factor ( sigma )
         if self.dim == 1:
             fac = self.fac * h1
         elif self.dim == 2:
             fac = self.fac * h1 * h1
         elif self.dim == 3:
             fac = self.fac * h1 * h1 * h1
 
-        # compute the gradient.
+        # compute sigma * dw_dq
         tmp2 = 2. - q
         if (rij > 1e-12):
             if (q > 2.0):
                 val = 0.0
             elif (q > 1.0):
-                val = -0.75 * tmp2 * tmp2 * h1 / rij
+                val = -0.75 * tmp2 * tmp2
             else:
-                val = -3.0 * q * (1 - 0.75 * q) * h1 / rij
+                val = -3.0 * q * (1 - 0.75 * q)
         else:
             val = 0.0
 
-        tmp = val * fac
+        return val * fac
+
+    def gradient(self, wdash=1., xij=[0., 0, 0], rij=1.0, h=1.0,
```

Comment:
Don't do this, as this means the gradient is not computed when you ask for it but requires wdash.  Instead compute wdash inside the gradient.